File: AutoJs6/src/core/Field.js
import { ColorChannel } from "../utils.js";

export default class Field {
  constructor(pxProfile) {
    this.pxProfile = pxProfile;
    this.imgs = {};
    this.isBossWave = false;
  }

  updateScreencaps(roiList, imgs) {
    for (let i = 0; i < roiList.length; i++) {
      const roi = roiList[i];
      const img = imgs[i];
      this.imgs[JSON.stringify(roi)] = img;
    }
  }

  clickLevelUp(deckIdx, times = 1) {
    for (let i = 0; i < times; i++) {
      // TODO
    }
  }

  clickEmote(emoteNo) {
    // TODO
  }

  waveProgressionDetected() {
    const roiKey = JSON.stringify(this.pxProfile.WAVE_CIRCLE_ROI);
    const img = this.imgs[roiKey];
    if (!img) {
      console.warn("ROI not found in imgs:", roiKey);
      return false;
    }

    const value = img[0][0][ColorChannel.R];
    const threshold = 255;
    const waveCircleIsLight = value >= threshold;

    let waveProgressionDetected = false;

    if (waveCircleIsLight) {
      if (!this.isBossWave) {
        this.isBossWave = true;
        waveProgressionDetected = true;
      }
    } else {
      this.isBossWave = false;
    }

    return waveProgressionDetected;
  }
}

File: PC/src/core/field.py
from .constants import *
from utils import ColorChannel
import numpy as np



class Field:
    def __init__(self, px_profile: PixelProfile):
        self.px_profile = px_profile

        self.imgs = dict()

        self.is_boss_wave = False

    def update_screencaps(self, roi_list, imgs: list[np.ndarray]):
        for (roi_list, img) in zip(roi_list, imgs):
            self.imgs[roi_list] = img

    def click_level_up(self, deck_idx, times=1):
        for i in range(times):
            pass

    def click_emote(self, emote_no):
        pass

    def wave_progression_detected(self):
        img = self.imgs.get(self.px_profile.WAVE_CIRCLE_ROI)
        if not img:
            return False
        value = img[0, 0, ColorChannel.R]
        threshold = 255
        wave_circle_is_light = value >= threshold

        wave_progression_detected = False

        if wave_circle_is_light:
            if not self.is_boss_wave:
                self.is_boss_wave = True
                wave_progression_detected = True
                # print("[wave] boss")
                # print(value)
        else:
            self.is_boss_wave = False

        return wave_progression_detected
